---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# phyloR

<!-- badges: start -->
```{r , echo=FALSE , results='asis' , message=FALSE}
cat(
badger::badge_devel("cparsania/tidywrappers" , color = "blue"),
badger::badge_lifecycle()
)

```
<!-- badges: end -->


```{r, echo=TRUE, message = FALSE, warning = FALSE , include=FALSE}
library(magrittr)
library(dplyr)

```

 `phyloR` is an R package to deal with NCBI-BLAST output. It helps to pre-process BLAST tabular output for downstream phylogenetic analysis.

## Install

```{r, echo=TRUE, message = FALSE, warning = FALSE,message=FALSE , eval=FALSE}
if(require("devtools")){
        devtools::install_github("cparsania/phyloR")
} else{
        install.packages("devtools")
        devtools::install_github("cparsania/phyloR")
}
```

## Assign column names to blast tabular output

```{r, echo=TRUE, message = FALSE, warning = FALSE,message=FALSE}

blast_out_file <- system.file("extdata" ,"blast_output_01.txt", package = "phyloR")
blast_out_tbl <- readr::read_delim(blast_out_file , delim = "\t" , comment = "#" ,col_names = F)

colnames(blast_out_tbl) <- phyloR::get_blast_outformat_7_colnames()
colnames(blast_out_tbl)
```

##  Filter blast hits

One can filter blast hits by individual or combinations of these variables.

+ e-value
+ bitscore
+ query coverage 
+ percent identity 

```{r, echo=TRUE, message = FALSE, warning = FALSE,message=FALSE}

## evalue <= 1e-5

blast_out_tbl %>% phyloR::filter_blast_hits(evalue = 1e-5)

##  Filter blast hits : bitscore >= 200

blast_out_tbl %>% phyloR::filter_blast_hits(bit_score = 200)

##  Filter blast hits : query coverage >= 90

blast_out_tbl %>% phyloR::filter_blast_hits(query_cov = 90, query_length = 253)

##  Filter blast hits : percent identity >= 90

blast_out_tbl %>% phyloR::filter_blast_hits(identity = 90)

##  Filter blast hits : evalue <= 1e-5, bitscore >= 200, query coverage >= 90, percent identity >= 60, 

blast_out_tbl %>% phyloR::filter_blast_hits(evalue = 1e-5, bit_score = 200, query_cov = 90, query_length = 253, identity = 60)
```

## Format fasta headers 

Fasta file downloaded as an output of blast result has fasta headers of two types 

1) Headers with alignment co-ordinates - if only aligned sequences downloaded. 

  + E.g. `KAE8371401.1:1-253 trypsin-like cysteine/serine peptidase domain-containing protein [Aspergillus bertholletius]`

2) Headers without alignment co-ordinates - if full sequences downloaded. 

  + E.g. `KAE8371401.1 trypsin-like cysteine/serine peptidase domain-containing protein [Aspergillus bertholletius]`

By default, `phyloR::format_fasta_headers()` divides headers in three (header type 2) or four (header type 1) groups delimited by 2 underscores (`__`). 

  1) subject id
  2) alignment coordinates (Optional for header type 1)
  3) subject description
  4) subject species 

One can drop alignment coordinates from sequence headers by keeping argument `keep_alignemnt_coord = FALSE`

```{r, echo=TRUE, message = FALSE, warning = FALSE,message=FALSE}

fa_file <- system.file("extdata" ,"blast_output_01.fasta", package = "phyloR")
fa <- Biostrings::readBStringSet(fa_file)

## existing headers 

names(fa) %>% head()

## new headers 

fa_file %>% phyloR::format_fasta_headers() %>% names() %>% head()

## drop alignment co-ordinates

fa_file %>% phyloR::format_fasta_headers(keep_alignemnt_coord = FALSE) %>% names() %>% head()

```

## Remove redundant hits from blast tabular ouput

BLAST reports same subject hit multiple time with different alignment co-ordinates. For a same subjet id, `phyloR::remove_redundant_hits()` select longest subject hit out of given multiples. 

```{r, echo=TRUE, message = FALSE, warning = FALSE}
blast_out_tbl %>% phyloR::remove_redundant_hits()
```

## Subset fasta sequences 

After filtering hits from tabular blast output, next obvious question is to filter same sequences from a fasta file and write them to new  file. `phyloR::subset_bstringset()` subset the sequences from a fasta using a vector of fasta headers. 

```{r}

fa_file <- system.file("extdata" ,"blast_output_01.fasta", package = "phyloR")
fa <- Biostrings::readBStringSet(fa_file)
query_headers <- fa %>% names() %>% sample(100)
seq_filtered <- phyloR::subset_bstringset(x = query_headers , y = fa, partial_match = F)

# One can use function `Biostrings::writeXStringSet()` to write filterd sequences to new fasta file. 
```


## Add taxonomy columns to blast tabular output

Adding taxonomy details to blast output is very crucial for downstream phylogenetic analysis. For example, a phylogenetic tree generated from sequence of blast output often required  to color tree branches either by kingdom, family or any other taxonomical level. `phyloR::add_taxonomy_columns()` adds user required taxonomy columns to blast tabular output. 

```{r, echo=TRUE, message = TRUE, warning = FALSE}

## add kingdom

with_kingdom <- blast_out_tbl %>% slice(1:10) %>% 
        phyloR::add_taxonomy_columns(ncbi_accession_colname = "subject_acc_ver", taxonomy_level = "kingdom")


with_kingdom %>% dplyr::select(query_acc_ver, subject_acc_ver, kingdom)

## add family  

with_kingdom_and_family <- with_kingdom %>% phyloR::add_taxonomy_columns(ncbi_accession_colname = "subject_acc_ver" ,
                                                                  taxonomy_level = "family")

with_kingdom_and_family %>% dplyr::select(query_acc_ver, subject_acc_ver, kingdom, family )

## add species 

with_kingdom_family_and_species <- with_kingdom_and_family %>% phyloR::add_taxonomy_columns(ncbi_accession_colname = "subject_acc_ver" ,
                                                                  taxonomy_level = "species")

with_kingdom_family_and_species %>% dplyr::select(query_acc_ver, subject_acc_ver, kingdom, family ,species)

```





