---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# phyloR

<!-- badges: start -->
```{r , echo=FALSE , results='asis' , message=FALSE}
cat(
badger::badge_devel("cparsania/tidywrappers" , color = "blue"),
badger::badge_lifecycle()
)

```
<!-- badges: end -->


```{r, echo=TRUE, message = FALSE, warning = FALSE , include=FALSE}
library(magrittr)
library(dplyr)

```

 `phyloR` is an R package. It helps to  pre-process the NCBI blast output for downstream phylogenetic analysis.

## Install

```{r, echo=TRUE, message = FALSE, warning = FALSE,message=FALSE , eval=FALSE}
if(require("devtools")){
        devtools::install_github("cparsania/phyloR")
} else{
        install.packages("devtools")
        devtools::install_github("cparsania/phyloR")
}
```

## Assign column names to blast tabular output

```{r, echo=TRUE, message = FALSE, warning = FALSE,message=FALSE}

blast_out_file <- system.file("extdata" ,"blast_output_01.txt", package = "phyloR")
blast_out_tbl <- readr::read_delim(blast_out_file , delim = "\t" , comment = "#" ,col_names = F)

colnames(blast_out_tbl) <- phyloR::get_blast_outformat_7_colnames()
colnames(blast_out_tbl)
```

##  Filter blast hits

```{r, echo=TRUE, message = FALSE, warning = FALSE,message=FALSE}

## evalue <= 1e-5

blast_out_tbl %>% phyloR::filter_blast_hits(evalue = 1e-5)

##  Filter blast hits : bitscore >= 200

blast_out_tbl %>% phyloR::filter_blast_hits(bit_score = 200)

##  Filter blast hits : query coverage >= 90

blast_out_tbl %>% phyloR::filter_blast_hits(query_cov = 90, query_length = 253)

##  Filter blast hits : percent identity >= 90

blast_out_tbl %>% phyloR::filter_blast_hits(identity = 90)

##  Filter blast hits : evalue <= 1e-5, bitscore >= 200, query coverage >= 90, percent identity >= 60, 

blast_out_tbl %>% phyloR::filter_blast_hits(evalue = 1e-5, bit_score = 200, query_cov = 90, query_length = 253, identity = 60)
```

## Format fasta headers 

```{r, echo=TRUE, message = FALSE, warning = FALSE,message=FALSE}

fa_file <- system.file("extdata" ,"blast_output_01.fasta", package = "phyloR")
fa <- Biostrings::readBStringSet(fa_file)

## existing headers 
names(fa) %>% head()

## new headers 
fa_file %>% phyloR::format_fasta_headers() %>% head()

```

## Remove redundancy from blast tabular ouput

```{r, echo=TRUE, message = FALSE, warning = FALSE}
blast_out_tbl %>% phyloR::remove_redundant_hits()
```

## Add taxonomy columns to blast tabular output

```{r, echo=TRUE, message = TRUE, warning = FALSE}

## add kingdom

with_kingdom <- blast_out_tbl %>% slice(1:10) %>% 
        phyloR::add_taxonomy_columns(ncbi_accession_colname = "subject_acc_ver", taxonomy_level = "kingdom")


with_kingdom

## add species 

with_kingdom_and_species <- with_kingdom %>% phyloR::add_taxonomy_columns(ncbi_accession_colname = "subject_acc_ver" ,
                                                                  taxonomy_level = "species")

with_kingdom_and_species
```





