<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clean blast results for downstream phyloegentic analysis.  • phyloR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Clean blast results for downstream phyloegentic analysis. ">
<meta property="og:description" content="phyloR is an R package to deal with NCBI-BLAST output. 
    It helps to pre-process BLAST tabular output for downstream phylogenetic analysis.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">phyloR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/cparsania/phyloR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="phylor" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#phylor" class="anchor"></a>phyloR</h1></div>
<!-- badges: start -->

<p><code>phyloR</code> is an R package to deal with NCBI-BLAST output. It helps to pre-process BLAST tabular output for downstream phylogenetic analysis.</p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<div class="sourceCode" id="cb1"><pre class="r"><span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="st">"devtools"</span>)){
        <span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"cparsania/phyloR"</span>)
} <span class="kw">else</span>{
        <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)
        <span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"cparsania/phyloR"</span>)
}</pre></div>
</div>
<div id="assign-column-names-to-blast-tabular-output" class="section level2">
<h2 class="hasAnchor">
<a href="#assign-column-names-to-blast-tabular-output" class="anchor"></a>Assign column names to blast tabular output</h2>
<div class="sourceCode" id="cb2"><pre class="r">
<span class="no">blast_out_file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span> ,<span class="st">"blast_output_01.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"phyloR"</span>)
<span class="no">blast_out_tbl</span> <span class="kw">&lt;-</span> <span class="kw pkg">readr</span><span class="kw ns">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_delim</a></span>(<span class="no">blast_out_file</span> , <span class="kw">delim</span> <span class="kw">=</span> <span class="st">"\t"</span> , <span class="kw">comment</span> <span class="kw">=</span> <span class="st">"#"</span> ,<span class="kw">col_names</span> <span class="kw">=</span> <span class="no">F</span>)

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">blast_out_tbl</span>) <span class="kw">&lt;-</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/get_blast_outformat_7_colnames.html">get_blast_outformat_7_colnames</a></span>()
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">blast_out_tbl</span>)
<span class="co">#&gt;  [1] "query_acc_ver"    "subject_acc_ver"  "identity"         "alignment_length"</span>
<span class="co">#&gt;  [5] "mismatches"       "gap_opens"        "q_start"          "q_end"           </span>
<span class="co">#&gt;  [9] "s_start"          "s_end"            "evalue"           "bit_score"       </span>
<span class="co">#&gt; [13] "positives"</span></pre></div>
</div>
<div id="filter-blast-hits" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-blast-hits" class="anchor"></a>Filter blast hits</h2>
<p>One can filter blast hits by individual or combinations of these variables.</p>
<ul>
<li>e-value</li>
<li>bitscore</li>
<li>query coverage</li>
<li>percent identity</li>
</ul>
<!-- end list --><div class="sourceCode" id="cb3"><pre class="r">
<span class="co">## evalue &lt;= 1e-5</span>

<span class="no">blast_out_tbl</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/filter_blast_hits.html">filter_blast_hits</a></span>(<span class="kw">evalue</span> <span class="kw">=</span> <span class="fl">1e-5</span>)
<span class="co">#&gt; # A tibble: 545 x 13</span>
<span class="co">#&gt;    query_acc_ver subject_acc_ver identity alignment_length mismatches gap_opens</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;              &lt;dbl&gt;            &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 KAE8371401.1  KAE8371401.1       100                253          0         0</span>
<span class="co">#&gt;  2 KAE8371401.1  XP_031924425.1      71.3              247         71         0</span>
<span class="co">#&gt;  3 KAE8371401.1  XP_031940570.1      66.4              247         83         0</span>
<span class="co">#&gt;  4 KAE8371401.1  KAB8261773.1        66.0              247         84         0</span>
<span class="co">#&gt;  5 KAE8371401.1  XP_022392557.1      64.4              247         69         1</span>
<span class="co">#&gt;  6 KAE8371401.1  OQD81666.1          61.7              248         95         0</span>
<span class="co">#&gt;  7 KAE8371401.1  GES63448.1          39.3              229        135         3</span>
<span class="co">#&gt;  8 KAE8371401.1  XP_001214927.1      39.3              229        135         3</span>
<span class="co">#&gt;  9 KAE8371401.1  OQD86157.1          38.6              228        132         3</span>
<span class="co">#&gt; 10 KAE8371401.1  XP_002375911.1      39.1              230        129         5</span>
<span class="co">#&gt; # … with 535 more rows, and 7 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;,</span>
<span class="co">#&gt; #   s_start &lt;dbl&gt;, s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;, positives &lt;dbl&gt;</span>

<span class="co">##  Filter blast hits : bitscore &gt;= 200</span>

<span class="no">blast_out_tbl</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/filter_blast_hits.html">filter_blast_hits</a></span>(<span class="kw">bit_score</span> <span class="kw">=</span> <span class="fl">200</span>)
<span class="co">#&gt; # A tibble: 6 x 13</span>
<span class="co">#&gt;   query_acc_ver subject_acc_ver identity alignment_length mismatches gap_opens</span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;              &lt;dbl&gt;            &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt; 1 KAE8371401.1  KAE8371401.1       100                253          0         0</span>
<span class="co">#&gt; 2 KAE8371401.1  XP_031924425.1      71.3              247         71         0</span>
<span class="co">#&gt; 3 KAE8371401.1  XP_031940570.1      66.4              247         83         0</span>
<span class="co">#&gt; 4 KAE8371401.1  KAB8261773.1        66.0              247         84         0</span>
<span class="co">#&gt; 5 KAE8371401.1  XP_022392557.1      64.4              247         69         1</span>
<span class="co">#&gt; 6 KAE8371401.1  OQD81666.1          61.7              248         95         0</span>
<span class="co">#&gt; # … with 7 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;, s_start &lt;dbl&gt;,</span>
<span class="co">#&gt; #   s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;, positives &lt;dbl&gt;</span>

<span class="co">##  Filter blast hits : query coverage &gt;= 90</span>

<span class="no">blast_out_tbl</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/filter_blast_hits.html">filter_blast_hits</a></span>(<span class="kw">query_cov</span> <span class="kw">=</span> <span class="fl">90</span>, <span class="kw">query_length</span> <span class="kw">=</span> <span class="fl">253</span>)
<span class="co">#&gt; # A tibble: 89 x 13</span>
<span class="co">#&gt;    query_acc_ver subject_acc_ver identity alignment_length mismatches gap_opens</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;              &lt;dbl&gt;            &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 KAE8371401.1  KAE8371401.1       100                253          0         0</span>
<span class="co">#&gt;  2 KAE8371401.1  XP_031924425.1      71.3              247         71         0</span>
<span class="co">#&gt;  3 KAE8371401.1  XP_031940570.1      66.4              247         83         0</span>
<span class="co">#&gt;  4 KAE8371401.1  KAB8261773.1        66.0              247         84         0</span>
<span class="co">#&gt;  5 KAE8371401.1  XP_022392557.1      64.4              247         69         1</span>
<span class="co">#&gt;  6 KAE8371401.1  OQD81666.1          61.7              248         95         0</span>
<span class="co">#&gt;  7 KAE8371401.1  GES63448.1          39.3              229        135         3</span>
<span class="co">#&gt;  8 KAE8371401.1  XP_001214927.1      39.3              229        135         3</span>
<span class="co">#&gt;  9 KAE8371401.1  XP_002375911.1      39.1              230        129         5</span>
<span class="co">#&gt; 10 KAE8371401.1  OQD72420.1          40                230        127         5</span>
<span class="co">#&gt; # … with 79 more rows, and 7 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;,</span>
<span class="co">#&gt; #   s_start &lt;dbl&gt;, s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;, positives &lt;dbl&gt;</span>

<span class="co">##  Filter blast hits : percent identity &gt;= 90</span>

<span class="no">blast_out_tbl</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/filter_blast_hits.html">filter_blast_hits</a></span>(<span class="kw">identity</span> <span class="kw">=</span> <span class="fl">90</span>)
<span class="co">#&gt; # A tibble: 1 x 13</span>
<span class="co">#&gt;   query_acc_ver subject_acc_ver identity alignment_length mismatches gap_opens</span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;              &lt;dbl&gt;            &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt; 1 KAE8371401.1  KAE8371401.1         100              253          0         0</span>
<span class="co">#&gt; # … with 7 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;, s_start &lt;dbl&gt;,</span>
<span class="co">#&gt; #   s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;, positives &lt;dbl&gt;</span>

<span class="co">##  Filter blast hits : evalue &lt;= 1e-5, bitscore &gt;= 200, query coverage &gt;= 90, percent identity &gt;= 60, </span>

<span class="no">blast_out_tbl</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/filter_blast_hits.html">filter_blast_hits</a></span>(<span class="kw">evalue</span> <span class="kw">=</span> <span class="fl">1e-5</span>, <span class="kw">bit_score</span> <span class="kw">=</span> <span class="fl">200</span>, <span class="kw">query_cov</span> <span class="kw">=</span> <span class="fl">90</span>, <span class="kw">query_length</span> <span class="kw">=</span> <span class="fl">253</span>, <span class="kw">identity</span> <span class="kw">=</span> <span class="fl">60</span>)
<span class="co">#&gt; # A tibble: 6 x 13</span>
<span class="co">#&gt;   query_acc_ver subject_acc_ver identity alignment_length mismatches gap_opens</span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;              &lt;dbl&gt;            &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt; 1 KAE8371401.1  KAE8371401.1       100                253          0         0</span>
<span class="co">#&gt; 2 KAE8371401.1  XP_031924425.1      71.3              247         71         0</span>
<span class="co">#&gt; 3 KAE8371401.1  XP_031940570.1      66.4              247         83         0</span>
<span class="co">#&gt; 4 KAE8371401.1  KAB8261773.1        66.0              247         84         0</span>
<span class="co">#&gt; 5 KAE8371401.1  XP_022392557.1      64.4              247         69         1</span>
<span class="co">#&gt; 6 KAE8371401.1  OQD81666.1          61.7              248         95         0</span>
<span class="co">#&gt; # … with 7 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;, s_start &lt;dbl&gt;,</span>
<span class="co">#&gt; #   s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;, positives &lt;dbl&gt;</span></pre></div>
</div>
<div id="format-fasta-headers" class="section level2">
<h2 class="hasAnchor">
<a href="#format-fasta-headers" class="anchor"></a>Format fasta headers</h2>
<p>Fasta file downloaded as an output of blast result has fasta headers of two types</p>
<p>1) Headers with alignment co-ordinates - if only aligned sequences downloaded.</p>
<!-- end list -->
<ul>
<li>E.g. <code>KAE8371401.1:1-253 trypsin-like cysteine/serine peptidase domain-containing protein [Aspergillus bertholletius]</code>
</li>
</ul>
<!-- end list --><p>2) Headers without alignment co-ordinates - if full sequences downloaded.</p>
<!-- end list -->
<ul>
<li>E.g. <code>KAE8371401.1 trypsin-like cysteine/serine peptidase domain-containing protein [Aspergillus bertholletius]</code>
</li>
</ul>
<p>By default, <code><a href="reference/format_fasta_headers.html">phyloR::format_fasta_headers()</a></code> divides headers in three (header type 2) or four (header type 1) groups delimited by 2 underscores (<code>__</code>).</p>
<p>1) subject id 2) alignment coordinates (Optional for header type 1) 3) subject description 4) subject species</p>
<p>One can drop alignment coordinates from sequence headers by keeping argument <code>keep_alignemnt_coord = FALSE</code></p>
<div class="sourceCode" id="cb4"><pre class="r">
<span class="no">fa_file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span> ,<span class="st">"blast_output_01.fasta"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"phyloR"</span>)
<span class="no">fa</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readBStringSet</a></span>(<span class="no">fa_file</span>)

<span class="co">## existing headers </span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">fa</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>()
<span class="co">#&gt; [1] "KAE8371401.1:1-253 trypsin-like cysteine/serine peptidase domain-containing protein [Aspergillus bertholletius]" </span>
<span class="co">#&gt; [2] "XP_031924425.1:1-247 trypsin-like cysteine/serine peptidase domain-containing protein [Aspergillus caelatus]"    </span>
<span class="co">#&gt; [3] "XP_031940570.1:1-247 trypsin-like cysteine/serine peptidase domain-containing protein [Aspergillus pseudonomius]"</span>
<span class="co">#&gt; [4] "KAB8261773.1:1-247 trypsin-like cysteine/serine peptidase domain-containing protein [Aspergillus pseudonomius]"  </span>
<span class="co">#&gt; [5] "XP_022392557.1:1-228 hypothetical protein ABOM_003016 [Aspergillus bombycis]"                                    </span>
<span class="co">#&gt; [6] "OQD81666.1:1-248 hypothetical protein PENANT_c026G03216 [Penicillium antarcticum]"</span>

<span class="co">## new headers </span>

<span class="no">fa_file</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/format_fasta_headers.html">format_fasta_headers</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>()
<span class="co">#&gt; [1] "KAE8371401_1__1_253__trypsin_like_cysteine_serine_peptidase_domain_containing_protein__Aspergillus_bertholletius" </span>
<span class="co">#&gt; [2] "XP_031924425_1__1_247__trypsin_like_cysteine_serine_peptidase_domain_containing_protein__Aspergillus_caelatus"    </span>
<span class="co">#&gt; [3] "XP_031940570_1__1_247__trypsin_like_cysteine_serine_peptidase_domain_containing_protein__Aspergillus_pseudonomius"</span>
<span class="co">#&gt; [4] "KAB8261773_1__1_247__trypsin_like_cysteine_serine_peptidase_domain_containing_protein__Aspergillus_pseudonomius"  </span>
<span class="co">#&gt; [5] "XP_022392557_1__1_228__hypothetical_protein_ABOM_003016__Aspergillus_bombycis"                                    </span>
<span class="co">#&gt; [6] "OQD81666_1__1_248__hypothetical_protein_PENANT_c026G03216__Penicillium_antarcticum"</span>

<span class="co">## drop alignment co-ordinates</span>

<span class="no">fa_file</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/format_fasta_headers.html">format_fasta_headers</a></span>(<span class="kw">keep_alignemnt_coord</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>()
<span class="co">#&gt; [1] "KAE8371401_1__trypsin_like_cysteine_serine_peptidase_domain_containing_protein__Aspergillus_bertholletius" </span>
<span class="co">#&gt; [2] "XP_031924425_1__trypsin_like_cysteine_serine_peptidase_domain_containing_protein__Aspergillus_caelatus"    </span>
<span class="co">#&gt; [3] "XP_031940570_1__trypsin_like_cysteine_serine_peptidase_domain_containing_protein__Aspergillus_pseudonomius"</span>
<span class="co">#&gt; [4] "KAB8261773_1__trypsin_like_cysteine_serine_peptidase_domain_containing_protein__Aspergillus_pseudonomius"  </span>
<span class="co">#&gt; [5] "XP_022392557_1__hypothetical_protein_ABOM_003016__Aspergillus_bombycis"                                    </span>
<span class="co">#&gt; [6] "OQD81666_1__hypothetical_protein_PENANT_c026G03216__Penicillium_antarcticum"</span></pre></div>
</div>
<div id="remove-redundant-hits-from-blast-tabular-ouput" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-redundant-hits-from-blast-tabular-ouput" class="anchor"></a>Remove redundant hits from blast tabular ouput</h2>
<p>BLAST reports same subject hit multiple time with different alignment co-ordinates. For a same subjet id, <code><a href="reference/remove_redundant_hits.html">phyloR::remove_redundant_hits()</a></code> select longest subject hit out of given multiples.</p>
<div class="sourceCode" id="cb5"><pre class="r"><span class="no">blast_out_tbl</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/remove_redundant_hits.html">remove_redundant_hits</a></span>()
<span class="co">#&gt; # A tibble: 716 x 13</span>
<span class="co">#&gt; # Groups:   subject_acc_ver [716]</span>
<span class="co">#&gt;    query_acc_ver subject_acc_ver identity alignment_length mismatches gap_opens</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;              &lt;dbl&gt;            &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 KAE8371401.1  1FN8_A              31.6              209        127         6</span>
<span class="co">#&gt;  2 KAE8371401.1  1GDU_A              31.9              210        127         6</span>
<span class="co">#&gt;  3 KAE8371401.1  1PPZ_A              31.9              210        127         6</span>
<span class="co">#&gt;  4 KAE8371401.1  AAQ04074.1          27.5              218        135         9</span>
<span class="co">#&gt;  5 KAE8371401.1  AAR91718.1          30.8              117         73         2</span>
<span class="co">#&gt;  6 KAE8371401.1  AAR91719.1          28.7              157         94         6</span>
<span class="co">#&gt;  7 KAE8371401.1  AAW31593.1          27.2              228        147         6</span>
<span class="co">#&gt;  8 KAE8371401.1  ADK37838.1          29.5              227        136         9</span>
<span class="co">#&gt;  9 KAE8371401.1  ADY16698.1          35.5              242        151         4</span>
<span class="co">#&gt; 10 KAE8371401.1  AER28315.1          30.5              236        148         5</span>
<span class="co">#&gt; # … with 706 more rows, and 7 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;,</span>
<span class="co">#&gt; #   s_start &lt;dbl&gt;, s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;, positives &lt;dbl&gt;</span></pre></div>
</div>
<div id="subset-fasta-sequences" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-fasta-sequences" class="anchor"></a>Subset fasta sequences</h2>
<p>After filtering hits from tabular blast output, next obvious question is to filter same sequences from a fasta file and write them to new file. <code><a href="reference/subset_bstringset.html">phyloR::subset_bstringset()</a></code> subset the sequences from a fasta using a vector of fasta headers.</p>
<div class="sourceCode" id="cb6"><pre class="r">
<span class="no">fa_file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span> ,<span class="st">"blast_output_01.fasta"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"phyloR"</span>)
<span class="no">fa</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readBStringSet</a></span>(<span class="no">fa_file</span>)
<span class="no">query_headers</span> <span class="kw">&lt;-</span> <span class="no">fa</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">100</span>)
<span class="no">seq_filtered</span> <span class="kw">&lt;-</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/subset_bstringset.html">subset_bstringset</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">query_headers</span> , <span class="kw">y</span> <span class="kw">=</span> <span class="no">fa</span>, <span class="kw">partial_match</span> <span class="kw">=</span> <span class="no">F</span>)

<span class="co"># One can use function `Biostrings::writeXStringSet()` to write filterd sequences to new fasta file. </span></pre></div>
</div>
<div id="add-taxonomy-columns-to-blast-tabular-output" class="section level2">
<h2 class="hasAnchor">
<a href="#add-taxonomy-columns-to-blast-tabular-output" class="anchor"></a>Add taxonomy columns to blast tabular output</h2>
<p>Adding taxonomy details to blast output is very crucial for downstream phylogenetic analysis. For example, a phylogenetic tree generated from sequence of blast output often required to color tree branches either by kingdom, family or any other taxonomical level. <code><a href="reference/add_taxonomy_columns.html">phyloR::add_taxonomy_columns()</a></code> adds user required taxonomy columns to blast tabular output.</p>
<div class="sourceCode" id="cb7"><pre class="r">
<span class="co">## add kingdom</span>

<span class="no">with_kingdom</span> <span class="kw">&lt;-</span> <span class="no">blast_out_tbl</span> <span class="kw">%&gt;%</span> <span class="fu">slice</span>(<span class="fl">1</span>:<span class="fl">10</span>) <span class="kw">%&gt;%</span>
        <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/add_taxonomy_columns.html">add_taxonomy_columns</a></span>(<span class="kw">ncbi_accession_colname</span> <span class="kw">=</span> <span class="st">"subject_acc_ver"</span>, <span class="kw">taxonomy_level</span> <span class="kw">=</span> <span class="st">"kingdom"</span>)
<span class="co">#&gt; No ENTREZ API key provided</span>
<span class="co">#&gt;  Get one via taxize::use_entrez()</span>
<span class="co">#&gt; See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/</span>
<span class="co">#&gt; ✓ Done.  Time taken 5.21</span>
<span class="co">#&gt; ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt; ● Rank search begins...</span>
<span class="co">#&gt; ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt; ✓ Done.  Time taken 0.08</span>


<span class="no">with_kingdom</span> <span class="kw">%&gt;%</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">query_acc_ver</span>, <span class="no">subject_acc_ver</span>, <span class="no">kingdom</span>)
<span class="co">#&gt; # A tibble: 10 x 3</span>
<span class="co">#&gt;    query_acc_ver subject_acc_ver kingdom</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;           &lt;chr&gt;  </span>
<span class="co">#&gt;  1 KAE8371401.1  KAE8371401.1    Fungi  </span>
<span class="co">#&gt;  2 KAE8371401.1  XP_031924425.1  Fungi  </span>
<span class="co">#&gt;  3 KAE8371401.1  XP_031940570.1  Fungi  </span>
<span class="co">#&gt;  4 KAE8371401.1  KAB8261773.1    Fungi  </span>
<span class="co">#&gt;  5 KAE8371401.1  XP_022392557.1  Fungi  </span>
<span class="co">#&gt;  6 KAE8371401.1  OQD81666.1      Fungi  </span>
<span class="co">#&gt;  7 KAE8371401.1  GES63448.1      Fungi  </span>
<span class="co">#&gt;  8 KAE8371401.1  XP_001214927.1  Fungi  </span>
<span class="co">#&gt;  9 KAE8371401.1  OQD86157.1      Fungi  </span>
<span class="co">#&gt; 10 KAE8371401.1  XP_002375911.1  Fungi</span>

<span class="co">## add family  </span>

<span class="no">with_kingdom_and_family</span> <span class="kw">&lt;-</span> <span class="no">with_kingdom</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/add_taxonomy_columns.html">add_taxonomy_columns</a></span>(<span class="kw">ncbi_accession_colname</span> <span class="kw">=</span> <span class="st">"subject_acc_ver"</span> ,
                                                                  <span class="kw">taxonomy_level</span> <span class="kw">=</span> <span class="st">"family"</span>)
<span class="co">#&gt; ── WARNING ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt; ℹ As column 'taxid' present in blast output tbl, same will be used to map taxonomy level.</span>
<span class="co">#&gt; ℹ To perform new 'taxid' search either remove  or rename columnn 'taxid'.</span>
<span class="co">#&gt; ── WARNING ENDS ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt; ● Rank search begins...</span>
<span class="co">#&gt; ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt; ✓ Done.  Time taken 0.05</span>

<span class="no">with_kingdom_and_family</span> <span class="kw">%&gt;%</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">query_acc_ver</span>, <span class="no">subject_acc_ver</span>, <span class="no">kingdom</span>, <span class="no">family</span> )
<span class="co">#&gt; # A tibble: 10 x 4</span>
<span class="co">#&gt;    query_acc_ver subject_acc_ver kingdom family        </span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;           &lt;chr&gt;   &lt;chr&gt;         </span>
<span class="co">#&gt;  1 KAE8371401.1  KAE8371401.1    Fungi   Aspergillaceae</span>
<span class="co">#&gt;  2 KAE8371401.1  XP_031924425.1  Fungi   Aspergillaceae</span>
<span class="co">#&gt;  3 KAE8371401.1  XP_031940570.1  Fungi   Aspergillaceae</span>
<span class="co">#&gt;  4 KAE8371401.1  KAB8261773.1    Fungi   Aspergillaceae</span>
<span class="co">#&gt;  5 KAE8371401.1  XP_022392557.1  Fungi   Aspergillaceae</span>
<span class="co">#&gt;  6 KAE8371401.1  OQD81666.1      Fungi   Aspergillaceae</span>
<span class="co">#&gt;  7 KAE8371401.1  GES63448.1      Fungi   Aspergillaceae</span>
<span class="co">#&gt;  8 KAE8371401.1  XP_001214927.1  Fungi   Aspergillaceae</span>
<span class="co">#&gt;  9 KAE8371401.1  OQD86157.1      Fungi   Aspergillaceae</span>
<span class="co">#&gt; 10 KAE8371401.1  XP_002375911.1  Fungi   Aspergillaceae</span>

<span class="co">## add species </span>

<span class="no">with_kingdom_family_and_species</span> <span class="kw">&lt;-</span> <span class="no">with_kingdom_and_family</span> <span class="kw">%&gt;%</span> <span class="kw pkg">phyloR</span><span class="kw ns">::</span><span class="fu"><a href="reference/add_taxonomy_columns.html">add_taxonomy_columns</a></span>(<span class="kw">ncbi_accession_colname</span> <span class="kw">=</span> <span class="st">"subject_acc_ver"</span> ,
                                                                  <span class="kw">taxonomy_level</span> <span class="kw">=</span> <span class="st">"species"</span>)
<span class="co">#&gt; ── WARNING ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt; ℹ As column 'taxid' present in blast output tbl, same will be used to map taxonomy level.</span>
<span class="co">#&gt; ℹ To perform new 'taxid' search either remove  or rename columnn 'taxid'.</span>
<span class="co">#&gt; ── WARNING ENDS ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt; ● Rank search begins...</span>
<span class="co">#&gt; ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt; ✓ Done.  Time taken 0.04</span>

<span class="no">with_kingdom_family_and_species</span> <span class="kw">%&gt;%</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">query_acc_ver</span>, <span class="no">subject_acc_ver</span>, <span class="no">kingdom</span>, <span class="no">family</span> ,<span class="no">species</span>)
<span class="co">#&gt; # A tibble: 10 x 5</span>
<span class="co">#&gt;    query_acc_ver subject_acc_ver kingdom family         species                 </span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;           &lt;chr&gt;   &lt;chr&gt;          &lt;chr&gt;                   </span>
<span class="co">#&gt;  1 KAE8371401.1  KAE8371401.1    Fungi   Aspergillaceae Aspergillus bertholleti…</span>
<span class="co">#&gt;  2 KAE8371401.1  XP_031924425.1  Fungi   Aspergillaceae Aspergillus caelatus    </span>
<span class="co">#&gt;  3 KAE8371401.1  XP_031940570.1  Fungi   Aspergillaceae Aspergillus pseudonomius</span>
<span class="co">#&gt;  4 KAE8371401.1  KAB8261773.1    Fungi   Aspergillaceae Aspergillus pseudonomius</span>
<span class="co">#&gt;  5 KAE8371401.1  XP_022392557.1  Fungi   Aspergillaceae Aspergillus bombycis    </span>
<span class="co">#&gt;  6 KAE8371401.1  OQD81666.1      Fungi   Aspergillaceae Penicillium antarcticum </span>
<span class="co">#&gt;  7 KAE8371401.1  GES63448.1      Fungi   Aspergillaceae Aspergillus terreus     </span>
<span class="co">#&gt;  8 KAE8371401.1  XP_001214927.1  Fungi   Aspergillaceae Aspergillus terreus     </span>
<span class="co">#&gt;  9 KAE8371401.1  OQD86157.1      Fungi   Aspergillaceae Penicillium antarcticum </span>
<span class="co">#&gt; 10 KAE8371401.1  XP_002375911.1  Fungi   Aspergillaceae Aspergillus flavus</span></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/cparsania/phyloR/">https://​github.com/​cparsania/​phyloR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/cparsania/phyloR/issues">https://​github.com/​cparsania/​phyloR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Chirag Parsania <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cparsania/tidywrappers"><img src="https://img.shields.io/badge/devel%20version-0.0.1.9000-blue.svg"></a></li>
<li><a href="https://www.tidyverse.org/lifecycle/#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Chirag Parsania.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
